src/go/overlord/statemanager.go

```diff
@@ -0,0 +1,168 @@
+//  Copyright 2018 Google LLC
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+
+//        https://www.apache.org/licenses/LICENSE-2.0
+
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//	limitations under the License.
+
+package overlord
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	mpb "github.com/google/minions/proto/minions"
+	pb "github.com/google/minions/proto/overlord"
+	"github.com/patrickmn/go-cache"
+)
+
+// StateManager handles the state of an Overlord through multiple
+// scans.
+type StateManager interface {
```

Comment:
Generally you want to define an interface at the place that uses that interface (in this case in overlord.go).

Reply:
Do you have a pointer for that? This is clearly Golang specific and I'm not familiar with the practice (shame on me). I'll gladly read up. This is effectively a very tightly coupled interface/implementations story, so I can see how that could work though.

Reply:
In go that is one of the applications of the [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle). With go type-system it translates to "accept interfaces return structs". So the package that creates the struct should return it as-is and the package who uses it should declare what it is using of it (often times using a narrow interface that does not require the whole struct contract to be passed).

Reply:
that makes sense. Thanks